File: app/src/main/java/com/example/sunddenfix/retrofit/utils/rx/RxConsumer.java
package com.example.sunddenfix.retrofit.utils.rx;

import android.util.Log;

import com.example.sunddenfix.retrofit.model.ActionResult;
import com.example.sunddenfix.retrofit.utils.StringUtil;

import io.reactivex.annotations.NonNull;
import io.reactivex.functions.Consumer;

/**
 * Created by wangchengm
 * on 2017/6/19.
 * 对consumer对封装
 */

public class RxConsumer<T> implements Consumer<ActionResult<T>> {

   private static final String TAG = "RxConsumer";

   @Override
   public void accept(@NonNull ActionResult<T> tActionResult) throws Exception {
      //在这里可以根据返回的不同code 做不同的事情 类似403这类型的错误code便可以统一处理
      Log.d(TAG, tActionResult.toString());
      switch (tActionResult.getCode()) {
         case ActionResult.RESULT_CODE_NO_LOGIN:
            //返回403 token失效
            break;
         case ActionResult.RESULT_CODE_NO_FOUND:
            //返回404  页面未找到
            break;
         case ActionResult.RESULT_CODE_SUCCESS:
            //返回200 请求成功
            _onSuccess(tActionResult.getData());
            break;
         default:
            if (!StringUtil.isNullOrEmpty(tActionResult.getMessage())) {
               //如果有返回的message
               _onError(tActionResult.getMessage());
            } else {
               _onError("网络链接出错");
            }
      }
   }

   //该方法在实例化的时候留给自己去实现
   public void _onSuccess(T t) {

   }

   //该方法在实例化的时候留给自己去实现
   public void _onError(String error) {

   }
}

File: app/src/main/java/com/example/sunddenfix/retrofit/utils/rx/RxConsumer.kt
package com.example.sunddenfix.retrofit.utils.rx

import com.example.sunddenfix.retrofit.model.ActionResult
import io.reactivex.functions.Consumer


/**
 * Created by wangchengm
 * on 2018/5/21.
 * 对consumer对封装
 * 对于 Subscriber 只需要onNext方法的时候 就可以使用Consumer
 */

open class RxConsumer<T> : Consumer<ActionResult<T>> {

    override fun accept(t: ActionResult<T>?) {
        when (t?.code) {
            ActionResult.RESULT_CODE_NO_LOGIN -> {
                //403 token失效
            }
            ActionResult.RESULT_CODE_NO_FOUND -> {
                //404页面没找到
            }
            ActionResult.RESULT_CODE_SUCCESS -> {
                //200 成功
                onSuccess(t.data)
            }
            else -> {
                if (t?.message.isNullOrEmpty()) {
                    onErrors("网络错误")
                } else {
                    onErrors(t?.message)
                }
            }
        }
    }

    open fun onSuccess(result: T?) {

    }

    open fun onErrors(error: String?) {

    }

}